crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1537 @@
+ * @param propq is a properties string
+ * @param suite-in specifies the preferred suite or NULL for a random choice
+ * @param suite is the chosen or random suite
+ * @param pub a random value of the appropriate length for sender public value
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param ciphertext buffer with random value of the appropriate length
+ * @param ciphertext_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_get_grease_value(OSSL_LIB_CTX *libctx, const char *propq,
+                                 OSSL_HPKE_SUITE *suite_in,
+                                 OSSL_HPKE_SUITE *suite,
+                                 unsigned char *enc,
+                                 size_t *enclen,
+                                 unsigned char *ct,
+                                 size_t ctlen)
+{
+    OSSL_HPKE_SUITE chosen;
+    size_t plen = 0;
+    const OSSL_HPKE_KEM_INFO *kem_info = NULL;
+    const OSSL_HPKE_AEAD_INFO *aead_info = NULL;
+
+    if (enc == NULL || !enclen
+        || ct == NULL || ctlen == 0 || suite == NULL)
+        return 0;
+    if (suite_in == NULL) {
+        /* choose a random suite */
+        if (hpke_random_suite(libctx, propq, &chosen) != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    } else {
+        chosen = *suite_in;
+    }
+    kem_info = ossl_HPKE_KEM_INFO_find_id(chosen.kem_id);
+    if (kem_info == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    aead_info = ossl_HPKE_AEAD_INFO_find_id(chosen.aead_id);
+    if (aead_info == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (hpke_suite_check(chosen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *suite = chosen;
+    /* make sure room for tag and one plaintext octet */
+    if (aead_info->taglen >= ctlen) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* publen */
+    plen = kem_info->Npk;
+    if (plen > *enclen) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (RAND_bytes_ex(libctx, enc, plen, OSSL_HPKE_RSTRENGTH) <= 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *enclen = plen;
+    /* if NIST curve chosen set 1st octet to 0x04 */
+    if (hpke_kem_id_nist_curve(chosen.kem_id) == 1) {
+        enc[0] = 0x04;
+    }
+    if (RAND_bytes_ex(libctx, ct, ctlen, OSSL_HPKE_RSTRENGTH) <= 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    return 1;
+err:
+    return 0;
+}
+
+/*
+ * @brief tell the caller how big the ciphertext will be
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who knows what'll be true in the future.
```

Comment:
replace 'll with  will (everywhere you have used it).

Reply:
sure